File: backend/app/models/knowledge_source.py
"""KnowledgeSource ORM model.

Represents a single source of knowledge (a document collection, a
database connection, a website, or an enterprise connector) that can
back one or more copilots.
"""

import uuid
from datetime import datetime, timezone

from sqlalchemy import DateTime, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database.base import Base
from app.models.copilot import copilot_knowledge_sources


class KnowledgeSource(Base):
    """A named source of knowledge documents (e.g. "HR Policies")."""

    __tablename__ = "knowledge_sources"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    # "documents" | "database" | "website" | "connector"
    source_type: Mapped[str] = mapped_column(String(50), nullable=False, default="documents")
    # "active" | "syncing" | "connected" | "pending" | "coming_soon"
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="active")

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    copilots: Mapped[list["Copilot"]] = relationship(
        "Copilot",
        secondary=copilot_knowledge_sources,
        back_populates="knowledge_sources",
    )
    documents: Mapped[list["Document"]] = relationship(
        "Document",
        back_populates="knowledge_source",
        cascade="all, delete-orphan",
    )

File: frontend/types/knowledge-source.ts
export type DocumentProcessingStatus = "UPLOADED" | "PROCESSING" | "READY" | "FAILED";
export type DocumentIndexStatus = "NOT_INDEXED" | "INDEXING" | "INDEXED" | "FAILED";

export interface ApiDocument {
  id: string;
  knowledge_source_id: string;
  name: string;
  status: "pending" | "processing" | "indexed";
  pages: number;
  chunks: number;
  embeddings: number;
  original_filename: string | null;
  processing_status: DocumentProcessingStatus | null;
  index_status: DocumentIndexStatus | null;
  created_at: string;
  updated_at: string;
}

export interface ApiKnowledgeSource {
  id: string;
  name: string;
  source_type: "documents" | "database" | "website" | "connector";
  status: "active" | "syncing" | "connected" | "pending" | "coming_soon";
  created_at: string;
  updated_at: string;
  documents: ApiDocument[];
}

export interface KnowledgeSourceCreatePayload {
  name: string;
  source_type?: ApiKnowledgeSource["source_type"];
  status?: ApiKnowledgeSource["status"];
}
